### PR TITLE
Unit testing with core 2 0

### LINKED ISSUE #2713: [Scenario] Unit testing for .NET Core/.NET Standard
Our tutorials show the basics of unit testing in .Net core projects and .NET standard projects.

We need to update those for .NET Standard 2.0.

One concern that we must handle is how testing a library whose only dependencies are .NET Standard libraries. Developers of these libraries will need to know how to manage CI builds and test environments that cover all the environments that are supported by these libraries.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #2713

Update C# unit testing with xUnit and MStest.
Add topics for VB and F# with both xUnit and MSTest.

### CODE REVIEW

File: docs/core/testing/unit-testing-fsharp-with-mstest.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        Source Files
+        MathService.fsproj
+    /MathService.Tests
+        Test Source Files
+        MathServiceTests.fsproj
+```
+
+## Creating the first test
+
+Before building the library or the tests, execute [`dotnet sln add .\PrimeService.Tests\PrimeService.Tests.csproj`](../tools/dotnet-restore.md) in the *unit-testing-using-dotnet-test* directory.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-fsharp-with-mstest.md]: .csproj --> .fsproj

File: docs/core/testing/unit-testing-fsharp-with-dotnet-test.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+```fsharp
+[<Fact>]
+let ``My test`` () =
+    Assert.True(true)
+
+[<Fact>]
+let ``Fail every time`` () = Assert.True(false)
+```
+
+The `[<Fact>]` attribute denotes a method as a single test. From the *unit-testing-with-fsharp*, execute [`dotnet test`](../tools/dotnet-test.md) to build the tests and the class library and then run the tests. The xUnit test runner contains the program entry point to run your tests. `dotnet test` starts the test runner using the unit test project you've created.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-fsharp-with-dotnet-test.md]: "denotes a method as a single test" is unclear -- I find "single test" misleading. The attributes denotes a particular method as a test method that's to be run by the test framework. Non-marked methods aren't run unless they're called from a test method. I'd change this to "denotes a test method that is be run by the test framework." This applies as well to the `[TestMethod]` attribute

File: docs/core/testing/unit-testing-fsharp-with-mstest.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
+---
+title: Unit testing F# libraries in .NET Core using dotnet test and xUnit
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-fsharp-with-mstest.md]: Not xUnit, but MSTest

File: docs/core/testing/unit-testing-fsharp-with-mstest.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+description: Learn unit test concepts for F# in .NET Core through an interactive experience building a sample solution step-by-step using dotnet test and xUnit.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-fsharp-with-mstest.md]: xUnit --> MSTest

File: docs/core/testing/unit-testing-fsharp-with-mstest.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+author: billwagner
+ms.author: wiwagn
+ms.date: 08/30/2017
+ms.topic: article
+ms.prod: .net-core
+---
+# Unit testing F# libraries in .NET Core using dotnet test and xUnit
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-fsharp-with-mstest.md]: xUnit --> MSTest

File: docs/core/testing/unit-testing-fsharp-with-mstest.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+```
+/unit-testing-with-fsharp
+    unit-testing-with-fsharp.sln
+    /MathService
+        Source Files
+        MathService.fsproj
+    /MathService.Tests
+```
+
+Make the *MathService.Tests* directory the current directory and create a new project using [`dotnet new mstest -lang F#`](../tools/dotnet-new.md). This creates a test project that uses xUnit as the test library. The generated template configures the test runner in the *MathServiceTests.fsproj*:
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-fsharp-with-mstest.md]: xUnit --> MSTest
test library, or test framework?

File: docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
+---
+title: Unit testing in .NET Core using dotnet test and xUnit
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md]: Title should also include VB

File: docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+description: Learn unit test concepts in .NET Core through an interactive experience building a sample visual basic solution step-by-step using dotnet test and xUnit.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md]: visual basic --> Visual Basic

File: docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+```
+/unit-testing-using-dotnet-test
+    unit-testing-using-dotnet-test.sln
+    /PrimeService
+```
+
+Make *PrimeService* the current directory and run [`dotnet new classlib -lang VB`](../tools/dotnet-new.md) to create the source project. Rename *Class1.VB* to *PrimeService.VB*. To use test-driven development (TDD), you create a failing implementation of the `PrimeService` class:
+
+```vb
+Imports System
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md]: You don't need to import System; it's imported automatically by the compiler.

File: docs/core/testing/unit-testing-visual-basic-with-mstest.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
+---
+title: Unit testing in .NET Core using dotnet test and MSTes
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-mstest.md]: MSTes --> MSTest
Add language to the title

File: docs/core/testing/unit-testing-visual-basic-with-mstest.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+description: Learn unit test concepts in .NET Core through an interactive experience building a sample visual basic solution step-by-step using MSTest.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-mstest.md]: visual basic --> Visual Basic

File: docs/core/testing/unit-testing-visual-basic-with-mstest.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+```xml
+<ItemGroup>
+<PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.3.0-preview-20170628-02" />
+<PackageReference Include="MSTest.TestAdapter" Version="1.1.18" />
+<PackageReference Include="MSTest.TestFramework" Version="1.1.18" />
+</ItemGroup>
+```
+
+The test project requires other packages to create and run unit tests. `dotnet new` in the previous step added xUnit and the xUnit runner. Now, add the `PrimeService` class library as another dependency to the project. Use the [`dotnet add reference`](../tools/dotnet-add-reference.md) command:
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-mstest.md]: not xUnit and xUnit runner.

File: docs/core/testing/unit-testing-visual-basic-with-mstest.md [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    End If
+    Throw New NotImplementedException("Please create a test first")
+End Function
+```
+
+In the *unit-testing-using-dotnet-test* directory, run `dotnet test` again. The `dotnet test` command runs a build for the `PrimeService` project and then for the `PrimeService.Tests` project. After building both projects, it runs this single test. It passes.
+
+## Adding more features
+
+Now that you've made one test pass, it's time to write more. There are a few other simple cases for prime numbers: 0, -1. You could add those cases as new tests with the `<TestMethod>` attribute, but that quickly becomes tedious. There are other xUnit attributes that enable you to write a suite of similar tests.  A `<<DataTestMethod>>` attribute represents a suite of tests that execute the same code but have different input arguments. You can use the `<DataRow>` attribute to specify values for those inputs.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-visual-basic-with-mstest.md]: `<<DataTestMethod>>` --> `<DataTestMethod>`

File: docs/core/testing/unit-testing-with-dotnet-test.md [hunk 14]
```diff
@@ -3,7 +3,7 @@ title: Unit testing in .NET Core using dotnet test and xUnit
 description: Learn unit test concepts in .NET Core through an interactive experience building a sample solution step-by-step using dotnet test and xUnit.
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-with-dotnet-test.md]: Title and description need to mention C#

File: docs/core/testing/unit-testing-with-mstest.md [hunk 15]
```diff
@@ -17,14 +17,16 @@ This tutorial takes you through an interactive experience building a sample solu
 
 ### Creating the source project
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-with-mstest.md]: The title and description should mention C#, and the description can be expanded

File: docs/core/testing/unit-testing-with-mstest.md [hunk 16]
```diff
@@ -43,23 +45,24 @@ namespace Prime.Services
...
 /unit-testing-using-mstest
+    unit-testing-using-mstest.sln
     /PrimeService
         Source Files
         PrimeService.csproj
     /PrimeService.Tests
 ```
 
-Make the *PrimeService.Tests* directory the current directory and create a new project using [`dotnet new mstest`](../tools/dotnet-new.md). This creates a test project that uses MStest as the test library. The generated template configures the test runner in the *PrimeServiceTests.csproj* file:
+Make the *PrimeService.Tests* directory the current directory and create a new project using [`dotnet new mstest`](../tools/dotnet-new.md). The Botnet new command creates a test project that uses MStest as the test library. The generated template configures the test runner in the *PrimeServiceTests.csproj* file:
```
Comment (rpetrusha) [inline docs/core/testing/unit-testing-with-mstest.md]: Botnet --> dotnet

File: samples/core/getting-started/unit-testing-vb-dotnet-test/PrimeService/PrimeService.vb [hunk 17]
```diff
@@ -0,0 +1,22 @@
+Imports System
```
Comment (rpetrusha) [inline samples/core/getting-started/unit-testing-vb-dotnet-test/PrimeService/PrimeService.vb]: No need to import System

File: samples/core/getting-started/unit-testing-vb-mstest/PrimeService/PrimeService.vb [hunk 18]
```diff
@@ -0,0 +1,22 @@
+Imports System
```
Comment (rpetrusha) [inline samples/core/getting-started/unit-testing-vb-mstest/PrimeService/PrimeService.vb]: No need to import System

**Review by rpetrusha — CHANGES_REQUESTED** [review]
Looks good, @BillWagner, despite a large number of comments.

File: docs/core/testing/index.md [hunk 19]
```diff
@@ -26,8 +30,20 @@ When writing unit tests, be careful you don’t accidentally introduce dependenc
...
+* Create unit tests using [*XUnit* and *C#* with the .NET Core CLI](unit-testing-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *C#* with the .NET Core CLI](unit-testing-with-mstest.md).
+* Create unit tests using [*XUnit* and *F#* with the .NET Core CLI](unit-testing-fsharp-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *F#* with the .NET Core CLI](unit-testing-fsharp-with-mstest.md).
+* Create unit tests using [*XUnit* and *Visual Basic* with the .NET Core CLI](unit-testing-visual-basic-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *Visual Basic* with the .NET Core CLI](unit-testing-visual-basic-with-mstest.md).
+
+You can choose different languages for your class libraries and your unit test libraries. You can learn how by mixing and matching the walk-throughs referenced above.
+
+* If you are using Visual Studio, checkout [Live Unit Testing in .NET Core]()
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/core/testing/index.md]: Thanks for spotting that I'd forgotten the link. (Yes, that's where I intended to go. I ran the samples, and the experience is the same for .NET Framework and .NET Core testing).

File: docs/core/testing/unit-testing-fsharp-with-dotnet-test.md [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+        Source Files
+        MathService.fsproj
+    /MathService.Tests
+        Test Source Files
+        MathServiceTests.fsproj
+```
+
+## Creating the first test
+
+Before building the library or the tests, execute [`dotnet sln add .\PrimeService.Tests\PrimeService.Tests.csproj`](../tools/dotnet-restore.md) in the *unit-testing-using-dotnet-test* directory.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/core/testing/unit-testing-fsharp-with-dotnet-test.md]: I was also linking to the wrong `dotnet` command.

File: docs/core/testing/unit-testing-fsharp-with-dotnet-test.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/core/testing/unit-testing-fsharp-with-dotnet-test.md]: I checked the others for similar errors.

File: docs/core/testing/index.md [hunk 21]
```diff
@@ -26,8 +30,20 @@ When writing unit tests, be careful you don’t accidentally introduce dependenc
 
 Learn more about unit testing in .NET Core projects:
 
-* Try this [walkthrough creating unit tests with xUnit and the .NET Core CLI](unit-testing-with-dotnet-test.md). 
+Unit Test projects for .NET Core are supported for [C#](../../csharp/index.md), [F#](../../fsharp/index.md) and [Visual Basic](../../visual-basic/index.md). You can also choose between [xUnit](http://xunit.github.io) and [MSTest](https://github.com/Microsoft/vstest-docs).
+
+You can read about those combinations in these walk-througs:
```
Comment (mairaw) [inline docs/core/testing/index.md]: typo: walk-througs. Also I think we don't hyphen that. So it should be walkthroughs.

File: docs/core/testing/index.md [hunk 22]
```diff
@@ -1,22 +1,26 @@
...
 ms.devlang: dotnet
 ms.assetid: 815ac74c-4bd9-4a94-a87c-78288b27c0e2
 ---
 
-# Unit Testing in .NET Core
+# Unit Testing in .NET Core and .NET Standard
 
 .NET Core has been designed with testability in mind, so that creating unit tests for your applications is easier than ever before. This article briefly introduces unit tests (and how they differ from other kinds of tests). Linked resources demonstrates how to add a test project to your solution and then run unit tests using either the command line or Visual Studio.
 
+.NET Core 2.0 supports [.NET Standard 2.0](../../standard/net-standard.md). The libraries used to demonstrate unit testing in this section rely on .NET Standard and will work in other project types as well.
```
Comment (mairaw) [inline docs/core/testing/index.md]: will work -> work

File: docs/core/testing/index.md [hunk 23]
```diff
@@ -1,22 +1,26 @@
...
+
+Beginning with .NET Core 2.0, templates for unit test project include VB.NET and F#, as well as C#.
+
 ## Getting Started with Testing
- 
+
 Having a suite of automated tests is one of the best ways to ensure a software application does what its authors intended it to do. There are many different kinds of tests for software applications, including integration tests, web tests, load tests, and many others. At the lowest level are unit tests, which test individual software components or methods. Unit tests should only test code within the developer’s control, and should not test infrastructure concerns, like databases, file systems, or network resources. Unit tests may be written using [Test Driven Development (TDD)](http://deviq.com/test-driven-development/), or they can be added to existing code to confirm its correctness. In either case, they should be small, well-named, and fast, since ideally you will want to be able to run hundreds of them before pushing your changes into the project’s shared code repository.
```
[the remark this answers is not in the document]
Comment (mairaw) [reply docs/core/testing/index.md]: Also the sentence that starts with "At the lowest level" seems to be missing a subject

will want -> want

File: docs/core/testing/index.md [hunk 24]
```diff
@@ -26,8 +30,20 @@ When writing unit tests, be careful you don’t accidentally introduce dependenc
...
+You can read about those combinations in these walk-througs:
+
+* Create unit tests using [*XUnit* and *C#* with the .NET Core CLI](unit-testing-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *C#* with the .NET Core CLI](unit-testing-with-mstest.md).
+* Create unit tests using [*XUnit* and *F#* with the .NET Core CLI](unit-testing-fsharp-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *F#* with the .NET Core CLI](unit-testing-fsharp-with-mstest.md).
+* Create unit tests using [*XUnit* and *Visual Basic* with the .NET Core CLI](unit-testing-visual-basic-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *Visual Basic* with the .NET Core CLI](unit-testing-visual-basic-with-mstest.md).
+
+You can choose different languages for your class libraries and your unit test libraries. You can learn how by mixing and matching the walk-throughs referenced above.
```
Comment (mairaw) [inline docs/core/testing/index.md]: walk-throughs referenced above -> walkthroughs previously referenced

File: docs/core/testing/index.md [hunk 25]
```diff
@@ -26,8 +30,20 @@ When writing unit tests, be careful you don’t accidentally introduce dependenc
...
+
+* If you are using Visual Studio, checkout [Live Unit Testing in .NET Core]()
+* For a additional information and examples on how to use selective unit test filtering, see [Running selective unit tests](selective-unit-tests.md).
```
[the remark this answers is not in the document]
Comment (mairaw) [reply docs/core/testing/index.md]: For a additional -> For additional

**Comment by BillWagner** [comment]
I just pushed all the changes addressing feedback. Ready for another go @rpetrusha

**Comment by rpetrusha** [comment]
This looks good, @BillWagner. The only unaddressed comment is one from Maira, from test/index.md:  
walk-throughs referenced above -> walkthroughs previously referenced (on line 44)
With that change, this is ready to merge.

**Review by rpetrusha — APPROVED** [review]

File: docs/core/testing/index.md [hunk 26]
```diff
@@ -26,8 +30,20 @@ When writing unit tests, be careful you don’t accidentally introduce dependenc
...
+* Create unit tests using [*XUnit* and *C#* with the .NET Core CLI](unit-testing-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *C#* with the .NET Core CLI](unit-testing-with-mstest.md).
+* Create unit tests using [*XUnit* and *F#* with the .NET Core CLI](unit-testing-fsharp-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *F#* with the .NET Core CLI](unit-testing-fsharp-with-mstest.md).
+* Create unit tests using [*XUnit* and *Visual Basic* with the .NET Core CLI](unit-testing-visual-basic-with-dotnet-test.md).
+* Create unit tests using [*MSTest* and *Visual Basic* with the .NET Core CLI](unit-testing-visual-basic-with-mstest.md).
+
+You can choose different languages for your class libraries and your unit test libraries. You can learn how by mixing and matching the walk-throughs referenced above.
+
+* If you are using Visual Studio, just see [Live Unit Testing in .NET Core](~/visualstudio/test/live-unit-testing)
```
Comment (mairaw) [inline docs/core/testing/index.md]: to cross link to the visualstudio docs, you need to remove the ~.  

`[Live Unit Testing in .NET Core](/visualstudio/test/live-unit-testing)`

File: docs/core/testing/index.md [hunk 27]
```diff
@@ -1,23 +1,27 @@
...
 
-# Unit Testing in .NET Core
+# Unit Testing in .NET Core and .NET Standard
 
-.NET Core has been designed with testability in mind, so that creating unit tests for your applications is easier than ever before. This article briefly introduces unit tests (and how they differ from other kinds of tests). Linked resources demonstrates how to add a test project to your solution and then run unit tests using either the command line or Visual Studio.
+.NET Core has been designed with testability in mind, so that creating unit tests for your applications is easier than ever before. This article briefly introduces unit tests (and how they differ from other kinds of tests). Linked resources demonstrate how to add a test project to your solution and then run unit tests using either the command line or Visual Studio.
+
+.NET Core 2.0 supports [.NET Standard 2.0](../../standard/net-standard.md). The libraries used to demonstrate unit testing in this section rely on .NET Standard and will work in other project types as well.
+
+Beginning with .NET Core 2.0, there are unit test project templates for VB.NET and F# as well as C#.
```
Comment (mairaw) [inline docs/core/testing/index.md]: one little nit thing: I think throughout the docs we usually refer to VB as Visual Basic. Should we be consistent on that?

**Pull request merged**